middleware/nethttp/abi.go

```diff
@@ -69,6 +75,43 @@ func (h *host) log(ctx context.Context, m api.Module, ptr, size uint32) {
 	h.logger(msg)
 }
 
+// readRequestHeader is the WebAssembly function export
+// "http.read_request_header", which writes a header value to memory, and
+// returns the count of bytes written.
+//
+// # Notes
+//
+//   - valueLimit is the limit of bytes to write, which means this can
+//     truncate unless it is the correct size.
+//   - a zero result is possible when the value is empty or the header doesn't
+//     exist.
+func (h *host) readRequestHeader(ctx context.Context, m api.Module,
+	namePtr, nameSize, valuePtr, valueLimit uint32) uint32 {
+	// TODO: maybe getRequestHeaderValueSize for to avoid pre-allocating or
+	// truncating, and also to tell the difference between no header and empty.
+	name := requireReadString(ctx, m.Memory(), "name", namePtr, nameSize)
+	r := requestStateFromContext(ctx).request
+	if values := r.Header.Values(name); len(values) == 0 {
+		return 0
+	} else {
+		value := values[0]
+		size := uint32(len(value))
+		if size > valueLimit {
+			size = valueLimit
+			value = value[0:valueLimit]
+		}
+		m.Memory().Write(ctx, valuePtr, []byte(value))
```

Comment:
Drive by note, `WriteString` to accept a string without copying into a fresh `[]byte` or unsafe shenanigans would probably be nice at some point

Reply:
hmm I thought that casting to a []byte doesn't copy in go? if you find a link ping me so I can correct my brain.

Reply:
Not sure if authoritative but here's link. It's mainly since strings are immutable while bytes aren't 

https://stackoverflow.com/a/59210739

Reply:
Escape analysis may let it remove the copy, but not sure if it's safe to rely on that. At least it's not obvious that it's not copying without digging into how the compiler inlines / analyzes

Reply:
thanks https://github.com/tetratelabs/wazero/issues/819 and feel free if bored to implement it. if not, I will by the next release.